File: backend/controller/authController.js
import User from "../models/user.js";
import bcrypt from "bcryptjs";
import jwt from "jsonwebtoken";
import crypto from "crypto";
import asyncHandler from "../utils/asyncHandler.js";

/* ================= REGISTER ================= */
export const registerUser = asyncHandler(async (req, res) => {
  const { name, email, password } = req.body;

  if (!name || !email || !password) {
    res.status(400);
    throw new Error("All fields required");
  }

  const exists = await User.findOne({ email });

  if (exists) {
    res.status(400);
    throw new Error("User already exists");
  }

  const hash = await bcrypt.hash(password, 10);

  const user = await User.create({
    name,
    email,
    password: hash,
  });

  res.status(201).json({
    message: "Registered Successfully",
    userId: user._id,
  });
});

/* ================= LOGIN ================= */
export const loginUser = asyncHandler(async (req, res) => {
  const { email, password } = req.body;

  const user = await User.findOne({ email });

  if (!user) {
    res.status(400);
    throw new Error("User not found");
  }

  const match = await bcrypt.compare(password, user.password);

  if (!match) {
    res.status(400);
    throw new Error("Wrong password");
  }

  const token = jwt.sign(
    { id: user._id },
    process.env.JWT_SECRET,
    { expiresIn: "1d" }
  );

  res.json({
    message: "Login Success",
    token,
    user: {
      id: user._id,
      name: user.name,
      email: user.email,
      role: user.role,
    },
  });
});

/* ================= FORGOT PASSWORD ================= */
export const forgotPassword = asyncHandler(async (req, res) => {
  const { email } = req.body;

  const user = await User.findOne({ email });

  if (!user) {
    res.status(404);
    throw new Error("User not found");
  }

  const token = crypto.randomBytes(20).toString("hex");

  user.resetToken = token;
  user.resetExpire = Date.now() + 15 * 60 * 1000;

  await user.save();

  res.json({
    message: "Reset token generated",
    token,
  });
});

File: backend/controller/auth_controller.py
"""Authentication controller."""
import secrets
import time

from config.db import get_db
from models.user import new_user_document
from utils.errors import AppError
from utils.security import hash_password, verify_password, create_token


async def register_user(name: str, email: str, password: str) -> dict:
    name = name.strip()
    email = email.lower().strip()

    if not name or not email or not password:
        raise AppError(400, "All fields required")
    if len(password) < 8:
        raise AppError(400, "Password must be at least 8 characters")

    db = get_db()
    exists = await db.users.find_one({"email": email})

    if exists:
        raise AppError(400, "User already exists")

    hashed = hash_password(password)
    doc = new_user_document(name, email, hashed)
    result = await db.users.insert_one(doc)

    return {"message": "Registered Successfully", "userId": str(result.inserted_id)}


async def login_user(email: str, password: str) -> dict:
    email = email.lower().strip()
    db = get_db()
    user = await db.users.find_one({"email": email})

    # Keep login failures intentionally generic so callers cannot distinguish
    # whether an email exists.
    if not user or not verify_password(password, user["password"]):
        raise AppError(401, "Invalid email or password")

    token = create_token(str(user["_id"]))

    return {
        "message": "Login Success",
        "token": token,
        "user": {
            "id": str(user["_id"]),
            "name": user["name"],
            "email": user["email"],
            "role": user.get("role", "user"),
        },
    }


async def forgot_password(email: str) -> dict:
    db = get_db()
    email = email.lower().strip()
    user = await db.users.find_one({"email": email})

    # Do not reveal whether an account exists.
    if not user:
        return {"message": "If the account exists, a reset request has been created"}

    token = secrets.token_urlsafe(32)

    await db.users.update_one(
        {"_id": user["_id"]},
        {"$set": {"resetToken": token, "resetExpire": int(time.time() * 1000) + 15 * 60 * 1000}},
    )

    # This endpoint still returns the token because this project has no email
    # provider wired in yet. It must be replaced by email delivery before production.
    return {"message": "Reset token generated", "token": token}
